pkg/internal/source/kind.go

```diff
@@ -104,14 +177,69 @@ func (ks *Kind) String() string {
 // WaitForSync implements SyncingSource to allow controllers to wait with starting
 // workers until the cache is synced.
 func (ks *Kind) WaitForSync(ctx context.Context) error {
+	if func() bool {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		return ks.started == nil
+	}() {
+		return fmt.Errorf("cannot wait for sync on an unstarted source")
+	}
+
 	select {
 	case err := <-ks.started:
-		return err
+		if err != nil {
+			// Try to stop the source to clean up any resources
+			// and collect any errors.
+			if stopErr := ks.Stop(); stopErr != nil {
+				err = kerrors.NewAggregate([]error{err, fmt.Errorf("failed to stop source: %w", stopErr)})
+			}
+
+			return err
+		}
+
+		return nil
 	case <-ctx.Done():
-		ks.startCancel()
 		if errors.Is(ctx.Err(), context.Canceled) {
 			return nil
 		}
 		return errors.New("timed out waiting for cache to be synced")
 	}
 }
+
+// Stop is internal and should be called only by the Controller to stop the Source. It should block until the
+// Source has stopped.
+// Stop can be called multiple times, and should be idempotent. If the Source is not running, Stop will prevent
+// the Source from starting in the future without error.
+func (ks *Kind) Stop() error {
+	ks.mu.Lock()
+	defer ks.mu.Unlock()
+	// Ensure that when we release the lock, we stop an in-process & future calls to Start().
+	ks.canceled = true
+
+	if ks.cancel != nil {
+		ks.cancel()
+	}
+	// If we haven't started yet, there's nothing to stop.
+	if ks.started == nil {
+		return nil
+	}
+
+	var errs []error
+
+	// Wait for the started channel to be closed.
+	ks.mu.Unlock()
+	if err := <-ks.started; err != nil {
+		errs = append(errs, err)
+	}
+	ks.mu.Lock()
+
+	// Remove all event handlers.
+	if ks.registration != nil {
+		if err := ks.informer.RemoveEventHandler(ks.registration); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return kerrors.NewAggregate(errs)
```

Comment:
```suggestion
	// Remove event handler.
	if ks.registration != nil {
		if err := ks.informer.RemoveEventHandler(ks.registration); err != nil {
			return err
		}
	}

	return nil
```

Reply:
The comment is indeed wrong, the `errs = append(errs, err)` however is used started above this section.

Reply:
Ups, sounds fine 😀